byway/src/main/webapp/recommendations.js

```diff
@@ -12,94 +12,61 @@
+const MAX_RECOMMENDATIONS = 1;
 
-// Contains google.maps.LatLng objects.
-// Used as a center point to search around a region.
+// Holds regions google.maps.LatLng objects
 let regions = [];
 
 /**
- * Initializes the webpage with a map and other google
- * services. Creates a route between two endpoints.
- * Note: Temp setup to mimic a route between two endpoints.
- * TODO: Link/adapt with Justine's third page, which handles
- * initializing the map and other google services.
- */
-function initServices() {
-  // Modified version of Justine's implementation
-  let directionsService = new google.maps.DirectionsService();
-  let directionsRenderer = new google.maps.DirectionsRenderer();
-  const start = new google.maps.LatLng(37.7699298, -122.4469157);
-  const end = new google.maps.LatLng(37.7683909618184, -122.51089453697205);
-  const mapOptions = {
-    zoom: 14,
-    center: start
-  }
-  addDestinations(start, end);
-  map = new google.maps.Map(document.getElementById('map'), mapOptions);
-  directionsRenderer.setMap(map);
-  document.getElementById("route").addEventListener("click", () => {
-    calcRoute(directionsService, directionsRenderer, start, end);
-  });
-  placesService = new google.maps.places.PlacesService(map);
-}
-
-/**
- * Add start and end points of the route to regions to search around.
- * Temporary set up to mimic adding several endpoints between legs.
- * @param {google.maps.LatLng} start coordinate of route's startpoint
- * @param {google.maps.LatLng} end coordinate of route's endpoint
+ * Creates round-trip route with waypoints that loads onto the map.
+ * Partitions the route into regions, used to load recommendations
+ * around. Orders the waypoints for efficiency and updates trip logistics.
  */
-function addDestinations(start, end) {
-  destinations.add(start);
-  destinations.add(end);
-  regions.push(...destinations);
-}
-
-/**
- * Creates a route between two points and loads onto the map.
- * Finds points along the path to load the regions Array.
- * Loads recommendations centered around these points.
- * @param {DirectionsService} directionsService finds directions
- * @param {DirectionsRenderer} directionsRenderer renders the route
- * @param {LatLng} start starting point location
- * @param {LatLng} end ending point location
- */
-function calcRoute(directionsService, directionsRenderer, start, end) {
+function calcMainRoute() {
   resetUserAlerts();
   const request = {
-      origin:  start,
-      destination: end,
-      travelMode: 'DRIVING'
+    origin:  start.name,
+    destination: end.name,
+    travelMode: 'DRIVING',
+    waypoints:  route.map(waypoint => ({location: waypoint.geometry.location})),
+    optimizeWaypoints: true
   };
+  addMainStopsToRegions();
   directionsService.route(request, function(result, status) {
     if (status == 'OK') {
       directionsRenderer.setDirections(result);
       findRegions(result);
+      document.getElementById("loading").style.visibility = 'visible';
```

Comment:
Is the delay between getting the data and loading it really long enough to need a loading indicator in the UI? Its okay if it is, just wondering

Reply:
A user can pick several interests, and depending on how many steps along a route are long enough, there can be several regions. Since there is a text search per region per interest, they can pile up. Also, the delay is currently 250ms per text search, so it may take a few seconds. I'm planning on reducing it, maybe to 100ms per query, but not sure where the threshold lies.

Also, this stuff was meant for another PR (since I split), so I removed this loading functionality.